### PR TITLE
C++ version charRNN, small modifications

### PR DESCRIPTION
**Pull request by mz24cn**
Official team member doesn't agree to make data default to be kNullOp (https://github.com/dmlc/MXNet.cpp/pull/54), I have to modify:
Executor* exe = RNN.SimpleBind(device, args_map);
to be
Executor* exe = RNN.SimpleBind(device, args_map, {}, {{"data", kNullOp}});

So the SimpleBind can not be so simple, for majority of ML tasks.

### CODE REVIEW

**Comment by mz24cn** [comment]
Sorry, I am unable to resolve conflicts.
qq 20170106232720

**Comment by conopt** [comment]
Hi @mz24cn, you need to merge locally in order to resolve the conflict (BTW, the pull request will be updated automatically when you push new commits, that means you don't need to create a new pull request if you modify the code)

**Comment by mz24cn** [comment]
https://github.com/dmlc/mxnet/pull/3485 leads to all examples (except mlp) in MXNet.CPP, such as alexnet.cpp, googlenet.cpp, etc., including charRNN.cpp, are INVALID now. MXOptimizerFree, MXOptimizerFindCreator and other two C API were removed. So, how to port the changes to MXNet.CPP to keep the backward compatibility?

**Comment by mz24cn** [comment]
@lx75249 @piiswrong

**Comment by conopt** [comment]
@mz24cn I'm porting Optimizer to mxnet.cpp, I guess it will be done in 2 days.

**Comment by mz24cn** [comment]
Not only Optimizer was removed, Operator are renamed, "_Plus" disappeared. NNVM breaks many industry code, besides introduces an unclear future. MXNet.cpp needs carefully investigate totally on the influence by NNVM, if it is not abandoned. @lx75249 @piiswrong

**Comment by mz24cn** [comment]
appveyor error message indicates libmxnet.lib file should be upgraded.

**Comment by mz24cn** [comment]
Hi @lx75249, I have done the windows verification work. Besides the **"#include <iterator>"** in **operator.hpp** is required by VS2015, **optimizer.hpp** should be modified. VS2015 compiler (VS Community 2015, version 14.0.25431.01 Update 3) seems new and release the std::pair object, which leads to the returned pointer is invalid when out of closure.

**Comment by mz24cn** [comment]
Now charRNN.cpp can properly run on Windows and Linux. Please consider merge this pull request. Thank you.

File: .travis.yml [hunk 1]
```diff
@@ -1,48 +1,48 @@
...
-  directories:
-    - ${HOME}/.cache/usr
-
-notifications:
-# Emails are sent to the committer's git-configured email address by default,
-  email:
-    on_success: change
-    on_failure: always
-  #slack: dmlc:NmroCzntCiWOuxUZpii40USd
+sudo: false
```
Comment (conopt) [inline .travis.yml]: Use UNIX line endings

**Comment by conopt** [comment]
It seems the encoding and line ending of many files changed. Please clean them up.

**Comment by mz24cn** [comment]
Two remained problems:
1, Ugly warings occur:
qq 20170112150148
2, The performance is slightly downgraded from 240s/eporch to 260s/eporch.

**Comment by mz24cn** [comment]
Hi @lx75249 , I cleaned up files and added data default to kNullOp. Now only 9 files updated.

File: include/mxnet-cpp/symbol.hpp [hunk 2]
```diff
@@ -246,6 +246,8 @@ void Symbol::InferExecutorArrays(
     auto iter_req = grad_req_type.find(arg_name);
     if (iter_req != grad_req_type.end()) {
       grad_reqs->push_back(iter_req->second);
+	} else if (arg_name == "data") {
```
Comment (conopt) [inline include/mxnet-cpp/symbol.hpp]: tabs -> spaces (you can keep tabs in charrnn example)

how about making all arg_names which end with "data" kNullOp?

**Comment by mz24cn** [comment]
okay, it's done.

**Comment by conopt** [comment]
there are still lint errors, please fix them by checking https://travis-ci.org/dmlc/MXNet.cpp/jobs/191323175#L465

**Comment by mz24cn** [comment]
now only charRNN.cpp tabs issue left, which is you allowed. thanks

**Comment by conopt** [comment]
Lint check still complains (https://travis-ci.org/dmlc/MXNet.cpp/jobs/191593609#L898), like line too long and braces issues. I think you can also change tabs to spaces by the way, so that I don't need to add special configs for your example. (You can see the checks in the end of this page after pushing changes, and when all the checks succeed, I will merge this pr)

**Comment by mz24cn** [comment]
you see?

**Pull request merged**